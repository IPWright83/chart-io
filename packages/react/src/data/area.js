export const areaData = [
    { month: new Date(2024, 0), sales: 118, revenue: 487, profit: 153 },
    { month: new Date(2024, 1), sales: 177, revenue: 279, profit: 293 },
    { month: new Date(2024, 2), sales: 328, revenue: 479, profit: 104 },
    { month: new Date(2024, 3), sales: 134, revenue: 447, profit: 57 },
    { month: new Date(2024, 4), sales: 171, revenue: 125, profit: -120 },
    { month: new Date(2024, 5), sales: 193, revenue: 245, profit: 450 },
    { month: new Date(2024, 6), sales: 455, revenue: 488, profit: 128 },
];

