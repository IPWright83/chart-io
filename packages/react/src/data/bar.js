const barData = [
    { year: 2011, value: 45 },
    { year: 2012, value: 47 },
    { year: 2013, value: 52 },
    { year: 2014, value: 70 },
    { year: 2015, value: 75 },
    { year: 2016, value: 78 },
];

export { barData };
