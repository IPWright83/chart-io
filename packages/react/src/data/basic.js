export const basicData = [
    { month: new Date(2024, 0), sales: 118, revenue: 487, profit: 153, target: 500 },
    { month: new Date(2024, 1), sales: 177, revenue: 279, profit: 293, target: 400 },
    { month: new Date(2024, 2), sales: 328, revenue: 479, profit: 104, target: 350 },
    { month: new Date(2024, 3), sales: 134, revenue: 447, profit: 57, target: 600 },
    { month: new Date(2024, 4), sales: 171, revenue: 125, profit: -120, target: 100 },
    { month: new Date(2024, 5), sales: 193, revenue: 245, profit: 450, target: 400 },
    { month: new Date(2024, 6), sales: 455, revenue: 488, profit: 128, target: 500 },
];

