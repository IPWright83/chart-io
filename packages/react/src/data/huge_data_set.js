let huge_data_set = [];

for (let i = 0; i < 20000; i++) {
    huge_data_set.push({
        "Units Sold": Math.random() * 10000,
        "Total Profit": Math.random() * 1600000,
    });
}

export { huge_data_set };
