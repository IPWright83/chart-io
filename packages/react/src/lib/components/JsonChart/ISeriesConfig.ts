import type { IColor } from "@chart-io/core";

export interface ISeriesConfig {
    lines?: {
        /**
         * Custom colors that the series should use instead of the theme
         */
        colors?: IColor[];
        /**
         * Should these series be interactive
         */
        interactive?: boolean;
        /**
         * The set of y fields to use to access the data for each plot
         */
        ys: Array<string>;
    };
    areas?: {
        /**
         * Custom colors that the series should use instead of the theme
         */
        colors?: IColor[];
        /**
         * Should these series be interactive
         */
        interactive?: boolean;
        /**
         * Should the areas be stacked?
         */
        stacked?: boolean;
        /**
         * The set of y fields to use to access the data for each plot
         */
        ys: Array<string>;
    };
    scatters?: {
        /**
         * Custom colors that the series should use instead of the theme
         */
        colors?: IColor[];
        /**
         * Should these series be interactive
         */
        interactive?: boolean;
        /**
         * Override the radius of the points
         */
        radius?: number;
        /**
         * The set of y fields to use to access the data for each plot
         */
        ys: Array<string>;
    };
    columns?: {
        /**
         * Custom colors that the series should use instead of the theme
         */
        colors?: IColor[];
        /**
         * Should these series be interactive
         */
        interactive?: boolean;
        /**
         * Should the columns be stacked?
         */
        stacked?: boolean;
        /**
         * Should the columns be grouped?
         */
        grouped?: boolean;
        /**
         * The set of y fields to use to access the data for each plot
         */
        ys: Array<string>;
    };
    bars?: {
        /**
         * Custom colors that the series should use instead of the theme
         */
        colors?: IColor[];
        /**
         * Should these series be interactive
         */
        interactive?: boolean;
        /**
         * Should the bars be stacked?
         */
        stacked?: boolean;
        /**
         * Should the bars be grouped?
         */
        grouped?: boolean;
        /**
         * The set of x fields to use to access the data for each plot
         */
        xs: Array<string>;
    };
}
