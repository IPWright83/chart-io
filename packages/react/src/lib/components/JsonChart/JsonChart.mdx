import { Story, Canvas, Meta } from "@storybook/blocks";

<Meta title="Components/JsonChart" />

# JsonChart Component

The `<JsonChart>` component allows us to define an XYChart with a configuration that is more easily serialized to a database by splitting out all the props into a large JSON object. Using this approach we can build most charts like so:

<Canvas>
    <Story id="components-jsonchart--line" />
</Canvas>
<br />

## Creating a JsonChart

To create a JsonChart you simply need to provide it with 2 simple props `config` and `data`. The `config` prop however contains multiple settings within, depending on the type of chart you want:

| Prop     | Type     | Note                                 |
| -------- | -------- | ------------------------------------ |
| `data`   | `IData`  | The data for the chart               |
| `config` | `object` | The full configuration for the chart |

Breaking down the `config` object further:

| Prop              | Type            | Note                                                                                   |
| ----------------- | --------------- | -------------------------------------------------------------------------------------- |
| `chart`           | `IXYChartProps` | All the chart propers (except data)                                                    |
| `series`          | `ISeriesConfig` | The config for the series (split by `lines`, `areas`, `scatters`, `columns` & `bars` ) |
| `series.lines`    | `object`        | The config for the line plots                                                          |
| `series.scatters` | `object`        | The config for the scatter plots                                                       |
| `axis`            | `object`        | The axis config                                                                        |
| `axis.x`          | `IXAxisProps`   | The config for the x-axis                                                              |
| `axis.y`          | `IXAxisProps`   | The config for the x-axis                                                              |

An example of the config for the above chart is:

```
const config = useMemo({
    {
        chart: {
            width: 800,
        },
        axis: {
            x: {
                fields: "Order Date",
            },
        },
        series: {
            lines: {
                ys: ["Total Cost", "Total Profit"],
            },
        },
    },
});
```

You should remember to memoize objects when dealing with large config objects to prevent unnecessary re-renders.

### Limitations:

It's worth noting that a few fields that you can specify within the configuration, you will not be able to serialize to a database. Examples of these are the event handlers (`onMouseOver`, `onMouseOut`, `onClick`) and also some of the more advanced properties such as the Axis `tickFormat` prop which requires a function.
