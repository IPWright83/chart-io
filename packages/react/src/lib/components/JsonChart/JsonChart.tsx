import React, { forwardRef } from "react";
import type { IData } from "@chart-io/core";

import { Areas, Bars, Columns, Lines, Scatters } from "../Plots";
import type { IXAxisProps, IYAxisProps } from "../Axis";
import { XAxis, YAxis } from "../Axis";
import { IChartRef } from "../Chart";
import type { IXYChartProps } from "../XYChart";
import { useArray } from "../../hooks";
import { XYChart } from "../XYChart";

import type { ISeriesConfig } from "./ISeriesConfig";

export interface IJsonChartProps {
    config: {
        /**
         * The config options for the main chart component
         */
        chart: Omit<IXYChartProps, "data">;
        /**
         * The config options for the series
         */
        series: ISeriesConfig;
        /**
         * The config options for the axis
         */
        axis: {
            /**
             * Configuration for the X-Axis
             */
            x?: IXAxisProps;
            /**
             * Configuration for the Y-Axis
             */
            y?: IYAxisProps;
        };
    };

    /**
     * The data for the chart
     */
    data?: IData;
}

/**
 * Represents an XYChart that can be configured via a serializable JSON string. It is worth noting
 * that some features (e.g. event handlers) will need to by manually applied as these can't be
 * serialised to JSON
 * @param  props        The set of React properties
 * @return              The JsonChart component
 */
export const JsonChart = forwardRef<IChartRef, IJsonChartProps>(({ config, data }: IJsonChartProps, ref) => {
    const { chart, series, axis } = config;
    const { lines, areas, scatters, bars, columns } = series;

    // Automatically determine the fields from the plots themselves
    const xPlotFields = bars?.xs;
    const yPlotFields = [...(lines?.ys ?? []), ...(areas?.ys ?? []), ...(scatters?.ys ?? []), ...(columns?.ys ?? [])];

    // Grab any explicitly set fields
    const xs = useArray(axis.x?.fields ?? xPlotFields);
    const ys = useArray(axis.y?.fields ?? yPlotFields);

    return (
        <XYChart ref={ref} data={data} {...chart}>
            {!!xs && <XAxis fields={xs} {...(axis.x ?? {})} />}
            {!!ys && <YAxis fields={ys} {...(axis.y ?? {})} />}
            {!!lines && <Lines x={xs[0]} ys={lines.ys} colors={lines.colors} interactive={lines.interactive} />}
            {!!areas && (
                <Areas
                    x={xs[0]}
                    ys={areas.ys}
                    colors={areas.colors}
                    interactive={areas.interactive}
                    stacked={areas.stacked}
                />
            )}
            {!!scatters && (
                <Scatters
                    x={xs[0]}
                    ys={scatters.ys}
                    colors={scatters.colors}
                    interactive={scatters.interactive}
                    radius={scatters.radius}
                />
            )}
            {!!columns && (
                <Columns
                    x={xs[0]}
                    ys={columns.ys}
                    colors={columns.colors}
                    interactive={columns.interactive}
                    stacked={columns.stacked}
                    grouped={columns.grouped}
                />
            )}
            {!!bars && (
                <Bars
                    xs={bars.xs}
                    y={ys[0]}
                    colors={bars.colors}
                    interactive={bars.interactive}
                    stacked={bars.stacked}
                    grouped={bars.grouped}
                />
            )}
        </XYChart>
    );
});

JsonChart.displayName = "JsonChart";
