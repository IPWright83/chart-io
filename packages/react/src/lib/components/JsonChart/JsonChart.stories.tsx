import React from "react";
import { JsonChart } from "./JsonChart";

import { sales_records_dataset } from "../../../data/sales_records_dataset";

export default {
    title: "Components/JsonChart",
    component: JsonChart,
    parameters: {
        chromatic: { delay: 300 },
    },
};

const JsonChartTemplate = (args) => <JsonChart config={args.config} data={sales_records_dataset} />;

export const Line = JsonChartTemplate.bind({});
Line.storyName = "Line";
Line.args = {
    config: {
        chart: {
            width: 800,
        },
        axis: {
            x: {
                fields: "Order Date",
            },
        },
        series: {
            lines: {
                ys: ["Total Cost", "Total Profit"],
            },
        },
    },
};

export const Area = JsonChartTemplate.bind({});
Area.storyName = "Area";
Area.args = {
    config: {
        chart: {
            width: 800,
            zoomBrush: "inline",
        },
        axis: {
            x: {
                fields: "Order Date",
            },
        },
        series: {
            areas: {
                ys: ["Total Cost", "Total Profit"],
                stacked: true,
            },
        },
    },
};

export const Scatter = JsonChartTemplate.bind({});
Scatter.storyName = "Scatter";
Scatter.args = {
    config: {
        chart: {
            width: 800,
        },
        axis: {
            x: {
                fields: "Order Date",
            },
        },
        series: {
            scatters: {
                ys: ["Total Cost", "Total Profit"],
                radius: 10,
            },
        },
    },
};

export const Columns = JsonChartTemplate.bind({});
Columns.storyName = "Columns";
Columns.args = {
    config: {
        chart: {
            width: 800,
        },
        axis: {
            x: {
                fields: "Item Type",
            },
        },
        series: {
            columns: {
                ys: ["Unit Price", "Unit Cost"],
                grouped: true,
            },
        },
    },
};

export const Bars = JsonChartTemplate.bind({});
Bars.storyName = "Bars";
Bars.args = {
    config: {
        chart: {
            width: 800,
        },
        axis: {
            y: {
                fields: "Item Type",
            },
            x: {
                fields: ["Unit Price", "Unit Cost"],
                aggregate: true,
            },
        },
        series: {
            bars: {
                xs: ["Unit Price", "Unit Cost"],
                stacked: true,
            },
        },
    },
};
