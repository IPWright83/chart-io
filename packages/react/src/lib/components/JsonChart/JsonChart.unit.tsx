import React from "react";
import { render } from "@testing-library/react";

import { sales_records_dataset } from "../../../data/sales_records_dataset";

import { JsonChart } from "./JsonChart";

describe("JsonChart", () => {
    it("should create empty template correctly", () => {
        const { asFragment } = render(
            <JsonChart
                data={sales_records_dataset}
                config={{
                    chart: {
                        id: "0000",
                        width: 800,
                        animationDuration: 0,
                    },
                    axis: {
                        x: {
                            fields: "Order Date",
                        },
                    },
                    series: {
                        scatters: {
                            ys: ["Total Cost", "Total Profit"],
                            radius: 10,
                        },
                    },
                }}
            />,
        );

        expect(asFragment()).toMatchSnapshot();
    });
});
