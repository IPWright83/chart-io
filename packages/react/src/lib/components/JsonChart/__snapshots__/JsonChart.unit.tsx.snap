// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsonChart should create empty template correctly 1`] = `
<DocumentFragment>
  <div
    style="display: inline-block;"
  >
    <svg
      class="chart-svg"
      height="500"
      style="background-color: rgb(255, 255, 255);"
      width="800"
    >
      <g
        class="g-event-receiver"
      >
        <rect
          class="chart-io event-receiver"
          height="440"
          style="fill: none; pointer-events: all;"
          transform="translate(30, 30)"
          width="740"
        />
      </g>
      <defs>
        <clippath
          id="clip-path-0000"
        >
          <rect
            height="440"
            width="740"
            x="30"
            y="30"
          />
        </clippath>
      </defs>
      <g
        class="g-axis"
      >
        <g
          transform="translate(0, 470)"
        >
          <g
            class="g-gridlines bottom"
            fill="none"
            font-family="sans-serif"
            font-size="10"
            style="color: rgb(211, 211, 211); stroke-opacity: 0.8; stroke-width: 1; pointer-events: none;"
            text-anchor="middle"
          >
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(90.35646708138931,0)"
            >
              <line
                stroke="currentColor"
                y2="-440"
              />
              <text
                dy="0.71em"
                fill="currentColor"
                y="3"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(195.37201918092276,0)"
            >
              <line
                stroke="currentColor"
                y2="-440"
              />
              <text
                dy="0.71em"
                fill="currentColor"
                y="3"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(300.6752851218248,0)"
            >
              <line
                stroke="currentColor"
                y2="-440"
              />
              <text
                dy="0.71em"
                fill="currentColor"
                y="3"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(405.69083722135815,0)"
            >
              <line
                stroke="currentColor"
                y2="-440"
              />
              <text
                dy="0.71em"
                fill="currentColor"
                y="3"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(510.7063893208917,0)"
            >
              <line
                stroke="currentColor"
                y2="-440"
              />
              <text
                dy="0.71em"
                fill="currentColor"
                y="3"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(615.7219414204251,0)"
            >
              <line
                stroke="currentColor"
                y2="-440"
              />
              <text
                dy="0.71em"
                fill="currentColor"
                y="3"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(721.025207361327,0)"
            >
              <line
                stroke="currentColor"
                y2="-440"
              />
              <text
                dy="0.71em"
                fill="currentColor"
                y="3"
              />
            </g>
          </g>
          <g
            class="chart-io axis axis-bottom"
            fill="none"
            font-family="sans-serif"
            font-size="10"
            style="user-select: none; color: rgb(51, 51, 51); stroke-opacity: 0.8; stroke-width: 1;"
            text-anchor="middle"
          >
            <path
              class="domain"
              stroke="currentColor"
            />
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(90.35646708138931,0)"
            >
              <line
                stroke="currentColor"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="currentColor"
                y="9"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(195.37201918092276,0)"
            >
              <line
                stroke="currentColor"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="currentColor"
                y="9"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(300.6752851218248,0)"
            >
              <line
                stroke="currentColor"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="currentColor"
                y="9"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(405.69083722135815,0)"
            >
              <line
                stroke="currentColor"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="currentColor"
                y="9"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(510.7063893208917,0)"
            >
              <line
                stroke="currentColor"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="currentColor"
                y="9"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(615.7219414204251,0)"
            >
              <line
                stroke="currentColor"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="currentColor"
                y="9"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(721.025207361327,0)"
            >
              <line
                stroke="currentColor"
                y2="6"
              />
              <text
                dy="0.71em"
                fill="currentColor"
                y="9"
              />
            </g>
          </g>
        </g>
      </g>
      <g
        class="g-axis"
      >
        <g
          transform="translate(30, 0)"
        >
          <g
            class="g-gridlines left"
            fill="none"
            font-family="sans-serif"
            font-size="10"
            style="color: rgb(211, 211, 211); stroke-opacity: 0.8; stroke-width: 1; pointer-events: none;"
            text-anchor="end"
          >
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(0,470.5)"
            >
              <line
                stroke="currentColor"
                x2="740"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-3"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(0,426.5)"
            >
              <line
                stroke="currentColor"
                x2="740"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-3"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(0,382.5)"
            >
              <line
                stroke="currentColor"
                x2="740"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-3"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(0,338.5)"
            >
              <line
                stroke="currentColor"
                x2="740"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-3"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(0,294.5)"
            >
              <line
                stroke="currentColor"
                x2="740"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-3"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(0,250.5)"
            >
              <line
                stroke="currentColor"
                x2="740"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-3"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(0,206.5)"
            >
              <line
                stroke="currentColor"
                x2="740"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-3"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(0,162.50000000000003)"
            >
              <line
                stroke="currentColor"
                x2="740"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-3"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(0,118.49999999999999)"
            >
              <line
                stroke="currentColor"
                x2="740"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-3"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(0,74.5)"
            >
              <line
                stroke="currentColor"
                x2="740"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-3"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(0,30.5)"
            >
              <line
                stroke="currentColor"
                x2="740"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-3"
              />
            </g>
          </g>
          <g
            class="chart-io axis axis-left"
            fill="none"
            font-family="sans-serif"
            font-size="10"
            style="user-select: none; color: rgb(51, 51, 51); stroke-opacity: 0.8; stroke-width: 1;"
            text-anchor="end"
          >
            <path
              class="domain"
              stroke="currentColor"
            />
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(0,470.5)"
            >
              <line
                stroke="currentColor"
                x2="-6"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-9"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(0,426.5)"
            >
              <line
                stroke="currentColor"
                x2="-6"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-9"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(0,382.5)"
            >
              <line
                stroke="currentColor"
                x2="-6"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-9"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(0,338.5)"
            >
              <line
                stroke="currentColor"
                x2="-6"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-9"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(0,294.5)"
            >
              <line
                stroke="currentColor"
                x2="-6"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-9"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(0,250.5)"
            >
              <line
                stroke="currentColor"
                x2="-6"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-9"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(0,206.5)"
            >
              <line
                stroke="currentColor"
                x2="-6"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-9"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(0,162.50000000000003)"
            >
              <line
                stroke="currentColor"
                x2="-6"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-9"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(0,118.49999999999999)"
            >
              <line
                stroke="currentColor"
                x2="-6"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-9"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(0,74.5)"
            >
              <line
                stroke="currentColor"
                x2="-6"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-9"
              />
            </g>
            <g
              class="tick"
              opacity="0.000001"
              transform="translate(0,30.5)"
            >
              <line
                stroke="currentColor"
                x2="-6"
              />
              <text
                dy="0.32em"
                fill="currentColor"
                x="-9"
              />
            </g>
          </g>
        </g>
      </g>
      <g
        class="g-plot scatter"
        clip-path="url(#clip-path-0000)"
      />
      <g
        class="g-plot scatter"
        clip-path="url(#clip-path-0000)"
      />
      <g
        class="g-markers"
      />
      <g
        class="g-droplines"
      />
      <g
        class="g-tooltip-overlay"
      />
      <g
        class="g-legend-overlay"
      >
        <foreignobject
          height="500"
          style="width: 100%; height: 100%; pointer-events: none;"
          width="800"
        >
          <div
            class="chart-io legend"
            style="border: thin solid #ccc; display: flex; flex-direction: column; flex-wrap: wrap; position: absolute; right: 10px; transform: translateY(-50%); top: 50%; max-width: 200px; max-height: 460px; padding: 8px; background: rgb(255, 255, 255); opacity: 1; overflow: hidden;"
          >
            <div
              class="chart-io legend-item"
              style="display: flex; flex-direction: row; user-select: none; pointer-events: none; color: rgb(0, 0, 0); font-size: 12px; font-family: sans-serif;"
            >
              <div
                class="chart-io indicator-circle"
                style="background: rgba(153, 193, 220, 0.7); width: 10px; height: 10px; margin-top: 4px; margin-left: 5px; margin-right: 5px; border-radius: 50%; flex-shrink: 0;"
              />
              <div
                class="chart-io legend-values"
                style="display: flex; flex-direction: row; justify-content: space-between; width: 100%; flex-grow: 1; flex-shrink: 1;"
              >
                <span
                  class="chart-io legend-series-name"
                  style="margin-right: 15px; white-space: nowrap; overflow: hidden; text-overflow: ellipsis; flex-grow: 1; flex-shrink: 1;"
                >
                  Total Cost
                </span>
              </div>
            </div>
            <div
              class="chart-io legend-item"
              style="display: flex; flex-direction: row; user-select: none; pointer-events: none; color: rgb(0, 0, 0); font-size: 12px; font-family: sans-serif;"
            >
              <div
                class="chart-io indicator-circle"
                style="background: rgba(252, 153, 142, 0.7); width: 10px; height: 10px; margin-top: 4px; margin-left: 5px; margin-right: 5px; border-radius: 50%; flex-shrink: 0;"
              />
              <div
                class="chart-io legend-values"
                style="display: flex; flex-direction: row; justify-content: space-between; width: 100%; flex-grow: 1; flex-shrink: 1;"
              >
                <span
                  class="chart-io legend-series-name"
                  style="margin-right: 15px; white-space: nowrap; overflow: hidden; text-overflow: ellipsis; flex-grow: 1; flex-shrink: 1;"
                >
                  Total Profit
                </span>
              </div>
            </div>
          </div>
        </foreignobject>
      </g>
    </svg>
  </div>
</DocumentFragment>
`;
